proposals/0000-unsaturated-type-families.rst

```diff
@@ -0,0 +1,720 @@
+
+The ``a -> @m b`` syntax is thus syntactic sugar for ``(->) @m a b``.
+This proposed syntax hints at the optionality of manually specifying the
+matchability, a point discussed later in this proposal.
+
+Matchability polymorphism
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The version of ``Map`` above can only be applied to type families
+(which have kind ``-> @U``) but not type constructors (which have kind
+``-> @M``). Since matchabilities are a first-class type, they can be
+quantified over, thus enabling polymorphism in the matchability of arrows.
+
+This way, ``Map`` can be defined to be *matchability polymorphic* (in
+its first argument) ::
+
+   type family Map (f :: (->) @m a b) (xs :: [a]) :: [b] where
+     Map _ '[]       = '[]
+     Map f (x ': xs) = f x ': Map f xs
+
+As a matter of convenience for writing matchability-polymorphic arrow
+kinds, we allow matchability variables (i.e type variables of kind
+``Matchability``) in ``@`` annotations, in addition to ``U`` and ``M``.
+Thus, ``Map`` can alternatively be written as ::
+
+   type family Map (f :: a -> @m b) (xs :: [a]) :: [b] where
+     Map _ '[]       = '[]
+     Map f (x ': xs) = f x ': Map f xs
+
+These two variants of ``Map`` support taking both ``Id`` (a type family) and
+``Maybe`` (a type constructor) as the first argument. The complete kind of
+``Map`` is ``forall (m :: Matchability) a b. (a -> @m b) -> @U [a] -> @U [b]``.
+
+In fact, as evidenced by the change suggested in the *Syntax* section, we
+can even allow matchabilities to be applications of type families, e.g:::
+
+  type family Flip (m :: Matchability) :: Matchability where
+    Alternate 'Matchable = 'Unmatchable
+    Alternate 'Unmatchable = 'Matchable
+
+  type ArrFlip (m :: Matchability) a b = a -> @(Flip m) b
+
+  -- F only accepts 'f's with a matchable arrow kind.
+  type family F (f :: ArrFlip 'Unmatchable i j) (a :: i) :: j where
+    F f a = f a
+
+We could optionally add reserved operators for ``-> @U`` and ``-> @M``, e.g
+``a $-> b`` and ``a |-> b`` respectively. These would have to be baked in
+because GHC would have to treat them like ``->`` and GHC's parser handles this
+operator in a special way
+(see `here <https://gitlab.haskell.org/ghc/ghc/issues/10056#note_157509>`_).
+
+.. _Inference:
+
+Meaning of an annotation-free ``->``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Even though this proposal introduces a way to annotate arrows, we do not want
+to force every single arrow (written or inferred) to be annotated with a
+matchability. Therefore, let's consider the meaning that we want to give to an
+annotation-free ``->``, what matchability should be given by default? We could
+just default to matchable in all arrow kinds and unmatchable elsewhere. However,
+some contexts seem to be particularly adapted to a matchability-polymorphic
+interpretation, while others require a more rigid interpretation, constraining
+pieces of code to only accept type constructors with matchable arrow kinds.
```

Comment:
My personal opinion is that we want to be dumb rather than smart with this sort of things. If someone wants a matchability-polymorphic function, let them write it so.

It's my general philosophy that we shouldn't try and infer polymorphic things.

You have a bit of a complication anyway, which is that the `->` in types' kind signatures should mean matchable, while the `->` in terms' type signatures should mean unmatchable. So you can't be completely dumb. But the simplest the better anyway.